core/storage/exploration/gae_models.py

```diff
@@ -225,6 +211,56 @@ def delete_multi(
                 commit_log_models)
             datastore_services.put_multi(commit_log_models)
 
+    @staticmethod
+    def convert_to_valid_dict(snapshot_dict):
+        """Replace invalid fields and values in the ExplorationModel dict.
+        Some old ExplorationModels can contain fields
+        and field values that are no longer supported and would cause
+        an exception when we try to reconstitute a ExplorationModel from
+        them. We need to remove or replace these fields and values.
+
+        Args:
+            snapshot_dict: dict. The content of the model. Some fields and field
+                values might no longer exist in the ExplorationModel
+                schema.
+
+        Returns:
+            dict. The content of the model. Only valid fields and values are
+            present.
+        """
+
+        # The nodes field is moved to collection_contents dict. We
+        # need to move the values from nodes field to collection_contents dict
+        # and delete nodes.
+        if 'skill_tags' in snapshot_dict:
+            del snapshot_dict['skill_tags']
+        if 'default_skin' in snapshot_dict:
+            del snapshot_dict['default_skin']
+        if 'skin_customizations' in snapshot_dict:
+            del snapshot_dict['skin_customizations']
+
+        return snapshot_dict
+
+    def _reconstitute(self, snapshot_dict):
```

Comment:
@vojtechjelinek is it always the case that for deprecated properties we just need to delete the items from the values dict during reconstitution? If so, it might be easier for us to generalize this by making the base versioned model support providing a list of deprecated fields (e.g. via a class-defined property). That helps avoid overriding what is essentially the implementation details of the base class, and avoids repeating the extra complexity for each case where we want to delete fields.

Presumably, the map reduce job is also potentially generalizable, but I think it's reasonable to keep the one-time job approach we're doing now rather than trying to figure that out.

My perspective here is that supporting removing properties isn't something that will end with #10049. Presumably we will want to delete other properties in the future, and this seems like a lot of boilerplate in order to do so. WDYT?

Reply:
I guess we can try to come up with a boilerplate model class, but I don't want to give it high priority (in the Data and stability team) and not do it in this PR.

Reply:
Added #12059